### PR TITLE
fix(rust,python): make python `schema_overrides` information available to the rust-side inference code when initialising from records/dicts

### LINKED ISSUE #12032: `schema_overrides` loses precision on `uint` too large for `float`
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
rec = [{'id': 9187643043065364481, 'x': 5068140580, 't': 1698177078347106},
 {'id': 9223671840084328451, 'x': 5068228167, 't': 1698177095146143},
 {'id': 9223671840084328454, 'x': 5068282103, 't': 1698177107196849},
 {'id': 9223671840084328456, 'x': 5068285636, 't': 1698177108249264},
 {'id': 9187643043065364490, 'x': 5068697684, 't': 1698177190426178},
 {'id': 9223671840084328460, 'x': 5068741942, 't': 1698177198374138},
 {'id': 9187643043065364496, 'x': 5068745389, 't': 1698177199158213},
 {'id': 9223671840084328467, 'x': 5068837145, 't': 1698177214483470},
 {'id': 9205657441574846485, 'x': 5069114269, 't': 1698177261953686},
 {'id': 9187643043065364505, 'x': 5069184024, 't': 1698177272423798}]
pl.from_records(rec, schema_overrides={"id": pl.UInt64})
```

### Log output

```shell
id	x	t
u64	i64	i64
9187643043065364480	5068140580	1698177078347106
9223671840084328448	5068228167	1698177095146143
9223671840084328448	5068282103	1698177107196849
9223671840084328448	5068285636	1698177108249264
9187643043065364480	5068697684	1698177190426178
9223671840084328448	5068741942	1698177198374138
9187643043065364480	5068745389	1698177199158213
9223671840084328448	5068837145	1698177214483470
9205657441574846464	5069114269	1698177261953686
9187643043065364480	5069184024	1698177272423798
```

### Issue description

`id` column loses precision

### Expected behavior

expect `id` column to be loaded without loss of precision

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.11
Index type:          UInt32
Platform:            Linux-5.15.0-1031-aws-x86_64-with-glibc2.35
Python:              3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.2.1
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.1.0
gevent:              22.10.2
matplotlib:          3.7.1
numpy:               1.23.3
openpyxl:            3.0.9
pandas:              2.0.0
pyarrow:             9.0.0
pydantic:            1.10.7
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          1.4.47
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #12032.

* We weren't passing information available in the user-provided "schema_overrides" param down into the Rust-side functions where the initial type inference occurs; consequently large UInt64 values would first get inferred as Float64 _before_ being cast back to UInt64, thereby losing accuracy.

* This PR enhances the Rust-side `read_dicts` and `finish_from_rows` methods, allowing them to integrate the additional/optional information so that they can establish the correct dtypes for overridden columns _without_ first mediating via an inferred type.

## Before

```python
import polars as pl
recs = [
    {"id":9187643043065364490, "misc":333},
    {"id":9223671840084328467, "misc":666},
    {"id":9187643043065364505, "misc":999},
]
df = pl.from_records( recs, schema_overrides={"id":pl.UInt64} )
```

```python
df
# shape: (3, 2)
# ┌─────────────────────┬──────┐
# │ id                  ┆ misc │
# │ ---                 ┆ ---  │
# │ u64                 ┆ i64  │
# ╞═════════════════════╪══════╡
# │ 9187643043065364480 ┆ 333  │   << all "id" values are wrong as
# │ 9223671840084328448 ┆ 666  │      they were cast back from f64
# │ 9187643043065364480 ┆ 999  │
# └─────────────────────┴──────┘
```

## After

```python
df
# shape: (3, 2)
# ┌─────────────────────┬──────┐
# │ id                  ┆ misc │
# │ ---                 ┆ ---  │
# │ u64                 ┆ i64  │
# ╞═════════════════════╪══════╡
# │ 9187643043065364490 ┆ 333  │   << all "id" values are now correct
# │ 9223671840084328467 ┆ 666  │      (avoided intermediate type/cast)
# │ 9187643043065364505 ┆ 999  │
# └─────────────────────┴──────┘
```
Note that if "schema_overrides" is _not_ provided, the column is loaded as Float64, so the user should see that they need to provide an explicit override, where applicable.

---

🚀 As well as fixing this edge case, we should also get a little extra performance when loading from dicts where only _partial_ type information is provided (via "schema_overrides"), as the effected columns will no longer trigger a post-init cast - they will be directly loaded as the correct type.

👀 Note: if the whole schema is given up-front (via the "schema" param) then none of these issues apply and everything is loaded correctly/efficiently.

### CODE REVIEW

File: py-polars/src/dataframe.rs [hunk 1]
```diff
@@ -50,14 +50,15 @@ impl PyDataFrame {
...
         infer_schema_length: Option<usize>,
-        schema_overwrite: Option<Schema>,
+        schema: Option<Schema>,
+        schema_overrides_by_idx: Option<PlHashMap<usize, DataType>>,
     ) -> PyResult<Self> {
         // Object builder must be registered, this is done on import.
-        let schema =
-            rows_to_schema_supertypes(&rows, infer_schema_length.map(|n| std::cmp::max(1, n)))
-                .map_err(PyPolarsErr::from)?;
+        let s = rows_to_schema_supertypes(&rows, infer_schema_length.map(|n| std::cmp::max(1, n)))
```
Comment (ritchie46) [inline py-polars/src/dataframe.rs:57]: Can we make this name a bit more clear. `inferred_schema`

File: py-polars/src/dataframe.rs [hunk 2]
```diff
@@ -50,14 +50,15 @@ impl PyDataFrame {
...
+            .map_err(PyPolarsErr::from)?;
+
         // Replace inferred nulls with boolean and erase scale from inferred decimals.
-        let fields = schema.iter_fields().map(|mut fld| match fld.data_type() {
+        let fields = s.iter_fields().map(|mut fld| match fld.data_type() {
```
Comment (ritchie46) [inline py-polars/src/dataframe.rs:61]: This code was already here, but I don't understand why we don't modify the data types in place. As we convert to `final_schema` yet again on line 73.

File: py-polars/src/dataframe.rs [hunk 3]
```diff
@@ -512,35 +523,41 @@ impl PyDataFrame {
...
         // If given, read dict fields in schema order.
         let mut schema_columns = PlIndexSet::new();
-        if let Some(schema) = &schema_overwrite {
-            schema_columns.extend(schema.0.iter_names().map(|n| n.to_string()))
+        if let Some(s) = &schema {
+            schema_columns.extend(s.0.iter_names().map(|n| n.to_string()))
         }
         let (rows, names) = dicts_to_rows(dicts, infer_schema_length, schema_columns)?;
+
+        let mut schema_overrides_by_idx: PlHashMap<usize, DataType> = PlHashMap::new();
```
Comment (ritchie46) [inline py-polars/src/dataframe.rs:547]: Can we make this `Vec<(usize, DataType)>`? This doesn't need a hash table as there is a single index/column and we just iterate over this.

**Review by ritchie46 — CHANGES_REQUESTED** [review]

File: py-polars/src/dataframe.rs (hunk 3 above)
Comment (alexander-beedie) [reply py-polars/src/dataframe.rs]: Shame on me - was  originally expecting to do name lookups; fixed! 🤦

File: py-polars/src/dataframe.rs [hunk 4]
```diff
@@ -50,42 +50,57 @@ impl PyDataFrame {
...
-            },
-            _ => fld,
-        });
-        let mut schema = Schema::from_iter(fields);
 
-        if let Some(schema_overwrite) = schema_overwrite {
-            for (i, (name, dtype)) in schema_overwrite.into_iter().enumerate() {
-                if let Some((name_, dtype_)) = schema.get_at_index_mut(i) {
+        // Replace inferred nulls with boolean and erase scale from inferred decimals.
+        let mut final_schema =
```
Comment (ritchie46) [inline py-polars/src/dataframe.rs:62]: If we add this method to `Schema` in `polars-core`:

```rust
    /// Iterates over mut references to the dtypes in this schema
    pub fn iter_dtypes_mut(&mut self) -> impl Iterator<Item = &mut DataType> + '_ + ExactSizeIterator {
        self.inner.iter_mut().map(|(_name, dtype)| dtype)
    }
```

We can do this without reallocating the schema.

**Review by ritchie46 — CHANGES_REQUESTED** [review]

File: py-polars/src/dataframe.rs (hunk 4 above)
Comment (alexander-beedie) [reply py-polars/src/dataframe.rs:62]: Will get it done ✌️

**Comment by alexander-beedie** [comment]
FYI: the CI/CD failure is unrelated -
```
try to setup sccache version: v0.6.0
sccache download from url: 
  https://github.com/mozilla/sccache/releases/download/v0.6.0/sccache-v0.6.0-x86_64-unknown-linux-musl.tar.gz
Error: Unexpected HTTP response: 404
```

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**